### PR TITLE
MSYS-656 - fixed dsc_script for WMF5

### PR DESCRIPTION
**Pull request by piyushawasthi**
Fixes https://github.com/chef/chef/issues/5588

Signed-off-by: piyushawasthi <piyush.awasthi@msystechnologies.com>

### Description

[fixed dsc_script for WMF5]

### Issues Resolved

[dsc_script idempotence not working?]

### CODE REVIEW

File: lib/chef/util/dsc/local_configuration_manager.rb [hunk 1]
```diff
@@ -78,6 +78,18 @@ def lcm_command_code(configuration_path, test_only_parameters)
 EOH
     end
 
+    def test_lcm_command_code
+      ps_version = Mixlib::ShellOut.new("powershell $PSVersionTable.PSVersion.Major").run_command
```
Comment (stuartpreston) [inline lib/chef/util/dsc/local_configuration_manager.rb]: *or* use the query helper that is already there: ```Chef::Platform.supported_powershell_version?(node, 5)```

File: lib/chef/util/dsc/local_configuration_manager.rb [hunk 2]
```diff
@@ -78,6 +78,18 @@ def lcm_command_code(configuration_path, test_only_parameters)
...
+      ps_version.error!
+      if ps_version.stdout.to_i > 4
+        <<-EOH
+$ProgressPreference = 'SilentlyContinue';Test-DscConfiguration -path #{@configuration_path}
```
Comment (btm) [inline lib/chef/util/dsc/local_configuration_manager.rb]: Does this mean we still run `Start-DscConfiguration` and then `Test-DscConfiguration` to test? Wouldn't `Start-DscConfiguration` without `-whatif` actually make a change? I think `test_only_parameters` need to something more like `test_command`.

File: lib/chef/util/dsc/local_configuration_manager.rb (hunk 1 above)
Comment (piyushawasthi) [reply lib/chef/util/dsc/local_configuration_manager.rb]: Thanks @stuartpreston,

File: lib/chef/util/dsc/local_configuration_manager.rb (hunk 2 above)
Comment (piyushawasthi) [reply lib/chef/util/dsc/local_configuration_manager.rb]: First we run `Test-DscConfiguration` then `Start-DscConfiguration`.

File: lib/chef/util/dsc/local_configuration_manager.rb [hunk 3]
```diff
@@ -72,10 +72,24 @@ def run_configuration_cmdlet(configuration_document, apply_configuration, shello
...
+      else
+        <<-EOH
+$ProgressPreference = 'SilentlyContinue';start-dscconfiguration -path #{@configuration_path} -wait -erroraction 'stop' -force #{test_command}
+EOH
+      end
+    end
+
+    def test_lcm_command_code
+      if Chef::Platform.supported_powershell_version?(@node, 5)
+        "test"
```
Comment (NimishaS) [inline lib/chef/util/dsc/local_configuration_manager.rb]: Can we return something more specific to Powershell version 5 e.g. `test_powershell_5`? Returning `test` looks vague here.

**Comment by btm** [comment]
@piyushawasthi I'd appreciate it if you would refactor this a little more.

`test_supported_powershell` should return just a boolean with one purpose, something like `powershell_version_gte_5?` or `powershell_version_greater_than_5?` The string for powershell 4 that is returned by `test_supported_powershell` should just be part of `lcm_command_code` which should be told if we're testing or not, then look up the version, and return the right string to run. (I Think. but that's why this needs to be simpler).

**Comment by piyushawasthi** [comment]
@btm - I have updated code as per comment, Please review.

**Comment by btm** [comment]
@piyushawasthi take a look at this commit I pushed, this is more what I had in mind: https://github.com/chef/chef/commit/ca92458609a6a475eb69f6b94aac3dcd71e6ecbe

The goal here is to take everything related to the command we run and have it be in the same method. having the `"-whatif; if (! $?) { exit 1 }"` string in `run_configuration_cmdlet` and the rest of the strings in `lcm_command_code` made it hard to follow.

Can you add a unit test or two around the new PS5 behavior too? If you cherry-pick my commit you can just stub `ps_version_gte_5?` and write unit tests around `lcm_command` as it is now self-contained. Will probably have to update other tests due to the refactor though.

File: lib/chef/util/dsc/local_configuration_manager.rb [hunk 4]
```diff
@@ -72,10 +70,22 @@ def run_configuration_cmdlet(configuration_document, apply_configuration, shello
...
-EOH
+    def lcm_command(apply_configuration)
+      @common_command_prefix = "$ProgressPreference = 'SilentlyContinue';"
+      ps4_base_command = "#{@common_command_prefix} Start-DscConfiguration -path #{@configuration_path} -wait -erroraction 'stop' -force"
+      apply_configuration ? ps4_base_command : test_lcm_configuration(ps4_base_command)
+    end
+
+    def test_lcm_configuration(ps4_base_command)
+      if ps_version_gte_5?
+        "#{@common_command_prefix} Test-DscConfiguration -path #{@configuration_path}"
```
Comment (NimishaS) [inline lib/chef/util/dsc/local_configuration_manager.rb]: We are using an instance variable `@common_command_prefix` here which is required to be set prior to calling this method. If we need to use this `test_lcm_configuration` method for any other command in future then we'll have to make sure to define `@common_command_prefix` otherwise the function will throw error for PS 5.
It would be good to pass `common_command_prefix` as an argument.

File: spec/unit/util/dsc/local_configuration_manager_spec.rb [hunk 5]
```diff
@@ -63,7 +63,35 @@
         let(:lcm_standard_error) { nil }
         let(:lcm_cmdlet_success) { true }
 
-        it "should successfully return resource information for normally formatted output when cmdlet the cmdlet succeeds" do
+        it "successfully return resource information for normally formatted output when cmdlet the cmdlet succeeds" do
```
Comment (NimishaS) [inline spec/unit/util/dsc/local_configuration_manager_spec.rb]: `successfully *returns* resource information for normally formatted output when the cmdlet succeeds.`

File: spec/unit/util/dsc/local_configuration_manager_spec.rb [hunk 6]
```diff
@@ -63,7 +63,35 @@
...
+          test_configuration_result = lcm.test_configuration("config", {})
+          expect(test_configuration_result.class).to be(Array)
+          expect(test_configuration_result.length).to be > 0
+          expect(Chef::Log).not_to receive(:warn)
+        end
+      end
+
+      context "returns successfully for PowerShell version 5" do
```
Comment (NimishaS) [inline spec/unit/util/dsc/local_configuration_manager_spec.rb]: Nitpicking: It's not clear what it returns successfully for PowerShell version 5. It would be a good idea to use:
```
context "when running on PowerShell version 5" do
```

File: spec/unit/util/dsc/local_configuration_manager_spec.rb [hunk 7]
```diff
@@ -63,7 +63,35 @@
...
+        let(:lcm_standard_output) { normal_lcm_output }
+        let(:lcm_standard_error) { nil }
+        let(:lcm_cmdlet_success) { true }
+
+        it "successfully return resource information for normally formatted output when cmdlet the cmdlet succeeds" do
```
Comment (NimishaS) [inline spec/unit/util/dsc/local_configuration_manager_spec.rb]: successfully `returns`.

File: spec/unit/util/dsc/local_configuration_manager_spec.rb [hunk 8]
```diff
@@ -63,7 +63,35 @@
...
+          allow(lcm).to receive(:ps_version_gte_5?).and_return(true)
+          test_configuration_result = lcm.test_configuration("config", {})
+          expect(test_configuration_result.class).to be(Array)
+          expect(test_configuration_result.length).to be > 0
+          expect(Chef::Log).not_to receive(:warn)
+        end
+      end
+
+      context "returns successfully for PowerShell version less than 5" do
```
Comment (NimishaS) [inline spec/unit/util/dsc/local_configuration_manager_spec.rb]: Nitpicking: It's not clear what it returns successfully for PowerShell version less than 5. It would be a good idea to use:
```
context "when running on PowerShell version less than 5" do
```

File: spec/unit/util/dsc/local_configuration_manager_spec.rb [hunk 9]
```diff
@@ -63,7 +63,35 @@
...
+        let(:lcm_standard_output) { normal_lcm_output }
+        let(:lcm_standard_error) { nil }
+        let(:lcm_cmdlet_success) { true }
+
+        it "successfully return resource information for normally formatted output when cmdlet the cmdlet succeeds" do
```
Comment (NimishaS) [inline spec/unit/util/dsc/local_configuration_manager_spec.rb]: successfully `returns`. Please fix this in all the specs below from where `should` has been removed.

File: spec/unit/util/dsc/local_configuration_manager_spec.rb [hunk 10]
```diff
@@ -114,23 +142,23 @@
...
         end
       end
     end
 
-    it "should identify a correctly formatted error message as a resource import failure" do
+    it "identify a correctly formatted error message as a resource import failure" do
       expect(lcm.send(:dsc_module_import_failure?, dsc_resource_import_failure_output)).to be(true)
     end
 
-    it "should not identify an incorrectly formatted error message as a resource import failure" do
```
Comment (NimishaS) [inline spec/unit/util/dsc/local_configuration_manager_spec.rb]: `does` not identify an incorrectly formatted error message as a resource import failure.

Applicable to the specs below as well.

File: lib/chef/util/dsc/local_configuration_manager.rb (hunk 4 above)
Comment (btm) [reply lib/chef/util/dsc/local_configuration_manager.rb]: ` test_lcm_configuration` is only called from `lcm_command`. I'd combine it back into `lcm_command` similar to how I had https://github.com/chef/chef/commit/ca92458609a6a475eb69f6b94aac3dcd71e6ecbe. That way you don't have to set `@common_command_prefix` or pass `ps4_base_command` to a method that only needs it half the time.

File: spec/unit/util/dsc/local_configuration_manager_spec.rb [hunk 11]
```diff
@@ -114,23 +142,23 @@
         let(:lcm_standard_error) { "Abort, Retry, Fail?" }
         let(:lcm_cmdlet_success) { false }
 
-        it "should log a warning" do
+        it "log a warning" do
```
Comment (btm) [inline spec/unit/util/dsc/local_configuration_manager_spec.rb]: *logs a warning

File: spec/unit/util/dsc/local_configuration_manager_spec.rb [hunk 12]
```diff
@@ -94,14 +122,14 @@
...
-        it "should log a warning if the message is formatted as expected when a resource import failure occurs" do
+        it "log a warning if the message is formatted as expected when a resource import failure occurs" do
           expect(Chef::Log).to receive(:warn).at_least(:once)
           expect(lcm).to receive(:dsc_module_import_failure?).and_call_original
           test_configuration_result = nil
           expect { test_configuration_result = lcm.test_configuration("config", {}) }.not_to raise_error
         end
 
-        it "should return a (possibly empty) array of ResourceInfo instances" do
+        it "return a (possibly empty) array of ResourceInfo instances" do
```
Comment (btm) [inline spec/unit/util/dsc/local_configuration_manager_spec.rb]: *returns

File: spec/unit/util/dsc/local_configuration_manager_spec.rb [hunk 13]
```diff
@@ -94,14 +122,14 @@
         let(:lcm_standard_error) { dsc_resource_import_failure_output }
         let(:lcm_cmdlet_success) { false }
 
-        it "should log a warning if the message is formatted as expected when a resource import failure occurs" do
+        it "log a warning if the message is formatted as expected when a resource import failure occurs" do
```
Comment (btm) [inline spec/unit/util/dsc/local_configuration_manager_spec.rb]: logs

File: spec/unit/util/dsc/local_configuration_manager_spec.rb [hunk 14]
```diff
@@ -80,7 +108,7 @@
           expect(lcm.send(:whatif_not_supported?, no_whatif_lcm_output)).to be_truthy
         end
 
-        it "should should return a (possibly empty) array of ResourceInfo instances" do
+        it "return a (possibly empty) array of ResourceInfo instances" do
```
Comment (btm) [inline spec/unit/util/dsc/local_configuration_manager_spec.rb]: returns

File: spec/unit/util/dsc/local_configuration_manager_spec.rb (hunk 6 above)
Comment (btm) [reply spec/unit/util/dsc/local_configuration_manager_spec.rb]: it seems like if you end up with a combined `lcm_command`, you would want a test somewhere that ensures you get `-whatif` on ps4 and `test-dscconfiguration` on ps5.
